### ISSUE TITLE
Full implementation of amounts as per hledger spec

### ISSUE THREAD

**Issue by darylwright** [jonestristand/hledger-parser#6]
**I'm submitting a ...**
  [ ] bug report
  [ * ] feature request
  [ ] question about the decisions made in the repository
  [ ] question about how to use this project

**Summary**

As per the hledger spec, [amounts](https://hledger.org/1.30/hledger.html#amounts) can be expressed with a great level of flexibility. Currently, `hledger-parser` only implements a subset of the amount formats supported by hledger. For features like the [commodity directive](https://hledger.org/1.30/hledger.html#commodity-directive) to make sense to implement, `hledger-parser` must implement support for the full range of amount formats allowable.

For instance, the following number formats should be valid, but currently are not:

- 1 000 000.55
- 1.000.000,00
- 1 000 030,00
- 1000 000 00 000.00
- ~~1,000,000 00~~
- ~~1.000.000 00~~
- 1.000,
- 1 000.
- 1,000.
- $-1.00
- -$1
- + $1
- $-&nbsp;&nbsp;&nbsp;&nbsp;1 ; it is only necessary to store one space since, "One or more spaces between the sign and the number are acceptable when parsing (but they won't be displayed in output):" [source](https://hledger.org/1.30/hledger.html#amounts)
- - $ 1
- $ - 1

~~The only exception to the above patterns I'd suggest is that a space is invalid as a decimal mark, unless it is followed by at least one digit. This removes any ambiguity when lexing and parsing content after the amount. With a currency format like `1,000 CAD`, we wouldn't be able to distinguish if the space was the decimal mark, and there's no "space" between the number and commodity text, or otherwise. Likewise with the amount followed by a comment (`1,000 ; inline comment`).~~ *Update: spaces are invalid as decimal marks, see [documentation](https://hledger.org/1.30/hledger.html#decimal-mark-directive).*

Additionally, the formatting of the amount must be stored alongside the number, dash, and symbol. This is crucial for being able to store the representation of amounts as per the original journal. This allows us to serialize and deserialize amounts deterministically.

To summarize, there are 4 main objectives to resolve this issue:

1) Support for space and period digit group marks
2) Support for ~~space and~~ comma decimal marks
3) Support for multiple layouts of the dash/plus
4) Storage of formatting information for amounts (we have to know how amounts are formatted at every step)
